1. Limpieza y transformación de los datos


1.1 Importar librería pandas y darle el diminutivo "pd"

In [1]:
import pandas as pd

1.2 Importar los datos ``'fbref_laliga21.csv'`` desde la ruta de tu google drive o traerlos directamente desde el ordenador y ver las 5 primeras filas del dataframe."

In [2]:
df=pd.read_csv("/content/drive/MyDrive/CURSO PYTHON/Tarea/fbref_laliga21 (3).csv")

In [3]:
df

Rk                                     Player  Nation   Pos  \
0      1                Sabit Abdulai\Sabit-Abdulai  gh GHA    MF   
1      2                  Marcos Acuña\Marcos-Acuna  ar ARG    DF   
2      3              Bobby Adekanye\Bobby-Adekanye  nl NED  FWMF   
3      4  Martin Agirregabiria\Martin-Agirregabiria  es ESP  DFMF   
4      5                  Joseph Aidoo\Joseph-Aidoo  gh GHA    DF   
..   ...                                        ...     ...   ...   
577  578                Joseba Zaldúa\Joseba-Zaldua  es ESP    DF   
578  579                  Oier Zarraga\Oier-Zarraga  es ESP    MF   
579  580                Igor Zubeldia\Igor-Zubeldia  es ESP    DF   
580  581          Martín Zubimendi\Martin-Zubimendi  es ESP    MF   
581  582            Martin Ødegaard\Martin-Odegaard  no NOR  MFFW   

             Squad  Age  Born  MP  Starts   Min  ...   xG  npxG   xA  npxG+xA  \
0           Getafe   21  1999   3       0    60  ...  0.0   0.0  0.0      0.0   
1          Sevilla   28  1991  30      26  2330  ...  1.1   1.1  2.8      4.0   
2            Cádiz   21  1999   3       0    36  ...  0.2   0.2  0.0      0.2   
3           Alavés   24  1996  26      16  1558  ...  0.2   0.2  0.9      1.1   
4       Celta Vigo   24  1995  25      14  1289  ...  0.8   0.8  0.6      1.4   
..             ...  ...   ...  ..     ...   ...  ...  ...   ...  ...      ...   
577  Real Sociedad   28  1992  12       6   639  ...  0.1   0.1  0.8      0.8   
578  Athletic Club   21  1999   5       1   100  ...  0.1   0.1  0.0      0.1   
579  Real Sociedad   23  1997  24      21  1959  ...  1.0   1.0  0.7      1.8   
580  Real Sociedad   21  1999  31      17  1882  ...  0.6   0.6  0.6      1.1   
581    Real Madrid   21  1998   7       3   234  ...  0.1   0.1  0.4      0.5   

     xG.1  xA.1  xG+xA  npxG.1  npxG+xA.1  Matches  
0    0.00  0.00   0.00    0.00       0.00  Matches  
1    0.04  0.11   0.15    0.04       0.15  Matches  
2    0.43  0.00   0.43    0.43       0.43  Matches  
3    0.01  0.05   0.06    0.01       0.06  Matches  
4    0.06  0.04   0.10    0.06       0.10  Matches  
..    ...   ...    ...     ...        ...      ...  
577  0.01  0.11   0.11    0.01       0.11  Matches  
578  0.08  0.04   0.13    0.08       0.13  Matches  
579  0.05  0.03   0.08    0.05       0.08  Matches  
580  0.03  0.03   0.05    0.03       0.05  Matches  
581  0.04  0.15   0.19    0.04       0.19  Matches  

[582 rows x 33 columns]

In [4]:
df.head()

Rk                                     Player  Nation   Pos       Squad  \
0   1                Sabit Abdulai\Sabit-Abdulai  gh GHA    MF      Getafe   
1   2                  Marcos Acuña\Marcos-Acuna  ar ARG    DF     Sevilla   
2   3              Bobby Adekanye\Bobby-Adekanye  nl NED  FWMF       Cádiz   
3   4  Martin Agirregabiria\Martin-Agirregabiria  es ESP  DFMF      Alavés   
4   5                  Joseph Aidoo\Joseph-Aidoo  gh GHA    DF  Celta Vigo   

   Age  Born  MP  Starts   Min  ...   xG  npxG   xA  npxG+xA  xG.1  xA.1  \
0   21  1999   3       0    60  ...  0.0   0.0  0.0      0.0  0.00  0.00   
1   28  1991  30      26  2330  ...  1.1   1.1  2.8      4.0  0.04  0.11   
2   21  1999   3       0    36  ...  0.2   0.2  0.0      0.2  0.43  0.00   
3   24  1996  26      16  1558  ...  0.2   0.2  0.9      1.1  0.01  0.05   
4   24  1995  25      14  1289  ...  0.8   0.8  0.6      1.4  0.06  0.04   

   xG+xA  npxG.1  npxG+xA.1  Matches  
0   0.00    0.00       0.00  Matches  
1   0.15    0.04       0.15  Matches  
2   0.43    0.43       0.43  Matches  
3   0.06    0.01       0.06  Matches  
4   0.10    0.06       0.10  Matches  

[5 rows x 33 columns]

"1.3. Obtener el tamaño del dataframe, cuantas filas y columnas tiene."

In [5]:
df.shape

(582, 33)

"1.4. Acceder al nombre de las  columnas."

In [6]:
df.columns

Index(['Rk', 'Player', 'Nation', 'Pos', 'Squad', 'Age', 'Born', 'MP', 'Starts',
       'Min', '90s', 'Gls', 'Ast', 'G-PK', 'PK', 'PKatt', 'CrdY', 'CrdR',
       'Gls.1', 'Ast.1', 'G+A', 'G-PK.1', 'G+A-PK', 'xG', 'npxG', 'xA',
       'npxG+xA', 'xG.1', 'xA.1', 'xG+xA', 'npxG.1', 'npxG+xA.1', 'Matches'],
      dtype='object')

"1.5.  Mirar el formato de las columnas, para comprobar si hay alguna columnas numérica que se encuentra en formato string."

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 582 entries, 0 to 581
Data columns (total 33 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Rk         582 non-null    int64  
 1   Player     582 non-null    object 
 2   Nation     582 non-null    object 
 3   Pos        582 non-null    object 
 4   Squad      582 non-null    object 
 5   Age        582 non-null    int64  
 6   Born       582 non-null    int64  
 7   MP         582 non-null    int64  
 8   Starts     582 non-null    int64  
 9   Min        582 non-null    int64  
 10  90s        582 non-null    float64
 11  Gls        582 non-null    int64  
 12  Ast        582 non-null    int64  
 13  G-PK       582 non-null    int64  
 14  PK         582 non-null    int64  
 15  PKatt      582 non-null    int64  
 16  CrdY       582 non-null    int64  
 17  CrdR       582 non-null    int64  
 18  Gls.1      582 non-null    float64
 19  Ast.1      582 non-null    float64
 20  G+A       

"1.6. Cuántos valores NAN tenemos, en caso de que hayan sustituido por cero y volver a mirar la cantidad de valores nan para comprobar que ya no tenemos."

In [8]:
df.isnull().sum()

Rk           0
Player       0
Nation       0
Pos          0
Squad        0
Age          0
Born         0
MP           0
Starts       0
Min          0
90s          0
Gls          0
Ast          0
G-PK         0
PK           0
PKatt        0
CrdY         0
CrdR         0
Gls.1        0
Ast.1        0
G+A          0
G-PK.1       0
G+A-PK       0
xG           2
npxG         2
xA           2
npxG+xA      2
xG.1         2
xA.1         2
xG+xA        2
npxG.1       2
npxG+xA.1    2
Matches      0
dtype: int64

In [9]:
df.fillna(0,inplace=True)

In [10]:
df.isnull().sum()

Rk           0
Player       0
Nation       0
Pos          0
Squad        0
Age          0
Born         0
MP           0
Starts       0
Min          0
90s          0
Gls          0
Ast          0
G-PK         0
PK           0
PKatt        0
CrdY         0
CrdR         0
Gls.1        0
Ast.1        0
G+A          0
G-PK.1       0
G+A-PK       0
xG           0
npxG         0
xA           0
npxG+xA      0
xG.1         0
xA.1         0
xG+xA        0
npxG.1       0
npxG+xA.1    0
Matches      0
dtype: int64

"1.7.  Eliminar la columna ``'Rk'`` y ``'matches'``. Recuerda guardar los cambios, después de eliminarlas realiza un .head() para ver el resultado."

In [11]:
df.drop(["Rk","Matches"] ,axis=1,inplace=True)

In [12]:
df.head()

Player  Nation   Pos       Squad  Age  \
0                Sabit Abdulai\Sabit-Abdulai  gh GHA    MF      Getafe   21   
1                  Marcos Acuña\Marcos-Acuna  ar ARG    DF     Sevilla   28   
2              Bobby Adekanye\Bobby-Adekanye  nl NED  FWMF       Cádiz   21   
3  Martin Agirregabiria\Martin-Agirregabiria  es ESP  DFMF      Alavés   24   
4                  Joseph Aidoo\Joseph-Aidoo  gh GHA    DF  Celta Vigo   24   

   Born  MP  Starts   Min   90s  ...  G+A-PK   xG  npxG   xA  npxG+xA  xG.1  \
0  1999   3       0    60   0.7  ...    0.00  0.0   0.0  0.0      0.0  0.00   
1  1991  30      26  2330  25.9  ...    0.12  1.1   1.1  2.8      4.0  0.04   
2  1999   3       0    36   0.4  ...    0.00  0.2   0.2  0.0      0.2  0.43   
3  1996  26      16  1558  17.3  ...    0.06  0.2   0.2  0.9      1.1  0.01   
4  1995  25      14  1289  14.3  ...    0.00  0.8   0.8  0.6      1.4  0.06   

   xA.1  xG+xA  npxG.1  npxG+xA.1  
0  0.00   0.00    0.00       0.00  
1  0.11   0.15    0.04       0.15  
2  0.00   0.43    0.43       0.43  
3  0.05   0.06    0.01       0.06  
4  0.04   0.10    0.06       0.10  

[5 rows x 31 columns]

"1.8. Solucionar el problema de  los nombres de los jugadores. Dejar el nombre en correcto estado."

In [13]:
df.Player.values

array(['Sabit Abdulai\\Sabit-Abdulai', 'Marcos Acuña\\Marcos-Acuna',
       'Bobby Adekanye\\Bobby-Adekanye',
       'Martin Agirregabiria\\Martin-Agirregabiria',
       'Joseph Aidoo\\Joseph-Aidoo', 'Carlos Akapo\\Carlos-Akapo',
       'Paul Akouokou\\Paul-Akouokou', 'Jony Álamo\\Jony-Alamo',
       'Jordi Alba\\Jordi-Alba', 'Raúl Albiol\\Raul-Albiol',
       'Paco Alcácer\\Paco-Alcacer', 'Pedro Alcalá\\Pedro-Alcala',
       'Rubén Alcaraz\\Ruben-Alcaraz', 'Iván Alejo\\Ivan-Alejo',
       'Carles Aleñá\\Carles-Alena', 'Carles Aleñá\\Carles-Alena',
       'Alfon\\Alfon', 'Sergio Álvarez\\Sergio-Alvarez',
       'Yeray Álvarez\\Yeray-Alvarez',
       'Emmanuel Amankwaa Akurugu\\Emmanuel-Amankwaa-Akurugu',
       'Jon Ander Garrido\\Jon-Ander-Garrido', 'Ángel\\Angel',
       'Miguel Ángel Rubio\\Miguel-Angel-Rubio', 'José Ángel\\Jose-Angel',
       'José Antonio Miranda\\Jose-Antonio-Miranda',
       'Aritz Arambarri\\Aritz-Arambarri',
       'Mauro Arambarri\\Mauro-Arambarri', 'Néstor A

In [14]:
df.Player.apply(lambda name:name.split("\\")[0])

0             Sabit Abdulai
1              Marcos Acuña
2            Bobby Adekanye
3      Martin Agirregabiria
4              Joseph Aidoo
               ...         
577           Joseba Zaldúa
578            Oier Zarraga
579           Igor Zubeldia
580        Martín Zubimendi
581         Martin Ødegaard
Name: Player, Length: 582, dtype: object

In [15]:
df["Player"]=df.Player.apply(lambda name:name.split("\\")[0])
df.head()

Player  Nation   Pos       Squad  Age  Born  MP  Starts  \
0         Sabit Abdulai  gh GHA    MF      Getafe   21  1999   3       0   
1          Marcos Acuña  ar ARG    DF     Sevilla   28  1991  30      26   
2        Bobby Adekanye  nl NED  FWMF       Cádiz   21  1999   3       0   
3  Martin Agirregabiria  es ESP  DFMF      Alavés   24  1996  26      16   
4          Joseph Aidoo  gh GHA    DF  Celta Vigo   24  1995  25      14   

    Min   90s  ...  G+A-PK   xG  npxG   xA  npxG+xA  xG.1  xA.1  xG+xA  \
0    60   0.7  ...    0.00  0.0   0.0  0.0      0.0  0.00  0.00   0.00   
1  2330  25.9  ...    0.12  1.1   1.1  2.8      4.0  0.04  0.11   0.15   
2    36   0.4  ...    0.00  0.2   0.2  0.0      0.2  0.43  0.00   0.43   
3  1558  17.3  ...    0.06  0.2   0.2  0.9      1.1  0.01  0.05   0.06   
4  1289  14.3  ...    0.00  0.8   0.8  0.6      1.4  0.06  0.04   0.10   

   npxG.1  npxG+xA.1  
0    0.00       0.00  
1    0.04       0.15  
2    0.43       0.43  
3    0.01       0.06  
4    0.06       0.10  

[5 rows x 31 columns]

 "1.9. Solucionar el problema de la nacionalidad. Quedarnos con la nacionalidad que está en mayúsculas. "

In [16]:
df.Nation.values


array(['gh GHA', 'ar ARG', 'nl NED', 'es ESP', 'gh GHA', 'gq EQG',
       'ci CIV', 'es ESP', 'es ESP', 'es ESP', 'es ESP', 'es ESP',
       'es ESP', 'es ESP', 'es ESP', 'es ESP', 'es ESP', 'es ESP',
       'es ESP', 'gh GHA', 'es ESP', 'es ESP', 'es ESP', 'es ESP',
       'gq EQG', 'es ESP', 'uy URU', 'mx MEX', 'uy URU', 'es ESP',
       'es ESP', 'es ESP', 'es ESP', 'es ESP', 'es ESP', 'es ESP',
       'ar ARG', 'ng NGA', 'co COL', 'es ESP', 'es ESP', 'es ESP',
       'es ESP', 'es ESP', 'mk MKD', 'es ESP', 'es ESP', 'es ESP',
       'es ESP', 'es ESP', 'es ESP', 'ar ARG', 'es ESP', 'es ESP',
       'es ESP', 'fr FRA', 'es ESP', 'es ESP', 'es ESP', 'es ESP',
       'es ESP', 'es ESP', 'es ESP', 'fr FRA', 'ma MAR', 'ar ARG',
       'dk DEN', 'rs SRB', 'es ESP', 'cl CHI', 'cm CMR', 'ar ARG',
       'es ESP', 'es ESP', 'hr CRO', 'ar ARG', 'es ESP', 'es ESP',
       'uy URU', 'es ESP', 'ar ARG', 'es ESP', 'es ESP', 'es ESP',
       'es ESP', 'es ESP', 'fr FRA', 'es ESP', 'es ESP', 'es E

In [17]:
df.Nation.apply(lambda name:name.split(" ")[1])

0      GHA
1      ARG
2      NED
3      ESP
4      GHA
      ... 
577    ESP
578    ESP
579    ESP
580    ESP
581    NOR
Name: Nation, Length: 582, dtype: object

In [18]:
df["Nation"]=df.Nation.apply(lambda name:name.split(" ")[1])

In [19]:
df.head()

Player Nation   Pos       Squad  Age  Born  MP  Starts   Min  \
0         Sabit Abdulai    GHA    MF      Getafe   21  1999   3       0    60   
1          Marcos Acuña    ARG    DF     Sevilla   28  1991  30      26  2330   
2        Bobby Adekanye    NED  FWMF       Cádiz   21  1999   3       0    36   
3  Martin Agirregabiria    ESP  DFMF      Alavés   24  1996  26      16  1558   
4          Joseph Aidoo    GHA    DF  Celta Vigo   24  1995  25      14  1289   

    90s  ...  G+A-PK   xG  npxG   xA  npxG+xA  xG.1  xA.1  xG+xA  npxG.1  \
0   0.7  ...    0.00  0.0   0.0  0.0      0.0  0.00  0.00   0.00    0.00   
1  25.9  ...    0.12  1.1   1.1  2.8      4.0  0.04  0.11   0.15    0.04   
2   0.4  ...    0.00  0.2   0.2  0.0      0.2  0.43  0.00   0.43    0.43   
3  17.3  ...    0.06  0.2   0.2  0.9      1.1  0.01  0.05   0.06    0.01   
4  14.3  ...    0.00  0.8   0.8  0.6      1.4  0.06  0.04   0.10    0.06   

   npxG+xA.1  
0       0.00  
1       0.15  
2       0.43  
3       0.06  
4       0.10  

[5 rows x 31 columns]

"1.10.  Solucionar el problema de las posiciones. Que cada jugador tenga unicamente una posicón."

In [20]:
df.Pos.value_counts()

DF      178
MF      130
FW       68
FWMF     64
MFFW     57
GK       40
MFDF     21
DFMF     17
DFFW      4
FWDF      2
GKMF      1
Name: Pos, dtype: int64

In [21]:
df['Pos'] = df['Pos'].apply(lambda x: x[:2])


In [22]:
df.Pos.value_counts().head()

MF    208
DF    199
FW    134
GK     41
Name: Pos, dtype: int64

In [23]:
df.head()

Player Nation Pos       Squad  Age  Born  MP  Starts   Min  \
0         Sabit Abdulai    GHA  MF      Getafe   21  1999   3       0    60   
1          Marcos Acuña    ARG  DF     Sevilla   28  1991  30      26  2330   
2        Bobby Adekanye    NED  FW       Cádiz   21  1999   3       0    36   
3  Martin Agirregabiria    ESP  DF      Alavés   24  1996  26      16  1558   
4          Joseph Aidoo    GHA  DF  Celta Vigo   24  1995  25      14  1289   

    90s  ...  G+A-PK   xG  npxG   xA  npxG+xA  xG.1  xA.1  xG+xA  npxG.1  \
0   0.7  ...    0.00  0.0   0.0  0.0      0.0  0.00  0.00   0.00    0.00   
1  25.9  ...    0.12  1.1   1.1  2.8      4.0  0.04  0.11   0.15    0.04   
2   0.4  ...    0.00  0.2   0.2  0.0      0.2  0.43  0.00   0.43    0.43   
3  17.3  ...    0.06  0.2   0.2  0.9      1.1  0.01  0.05   0.06    0.01   
4  14.3  ...    0.00  0.8   0.8  0.6      1.4  0.06  0.04   0.10    0.06   

   npxG+xA.1  
0       0.00  
1       0.15  
2       0.43  
3       0.06  
4       0.10  

[5 rows x 31 columns]

 "1.11. Renombrar las columnas que contiene el valor .1, como por ejemplo ``'Ast.1'`` estas son columnas que pertenecen a las métricas por 90 min. Sustituir ``'Ast.1'`` por ``'Ast/90'``, hacer los mismo con las columnas que presentan este patrón"

In [24]:
df.columns


Index(['Player', 'Nation', 'Pos', 'Squad', 'Age', 'Born', 'MP', 'Starts',
       'Min', '90s', 'Gls', 'Ast', 'G-PK', 'PK', 'PKatt', 'CrdY', 'CrdR',
       'Gls.1', 'Ast.1', 'G+A', 'G-PK.1', 'G+A-PK', 'xG', 'npxG', 'xA',
       'npxG+xA', 'xG.1', 'xA.1', 'xG+xA', 'npxG.1', 'npxG+xA.1'],
      dtype='object')

In [25]:
df.rename(columns={'Gls.1':'Gls/90','Ast.1':'Ast/90', 'G-PK.1':'G-PK/90','xG.1':'xG/90','xA.1':'xA/90','npxG.1':'npxG/90','npxG+xA.1':'npxG+xA/90'},inplace=True)


In [26]:
df.columns


Index(['Player', 'Nation', 'Pos', 'Squad', 'Age', 'Born', 'MP', 'Starts',
       'Min', '90s', 'Gls', 'Ast', 'G-PK', 'PK', 'PKatt', 'CrdY', 'CrdR',
       'Gls/90', 'Ast/90', 'G+A', 'G-PK/90', 'G+A-PK', 'xG', 'npxG', 'xA',
       'npxG+xA', 'xG/90', 'xA/90', 'xG+xA', 'npxG/90', 'npxG+xA/90'],
      dtype='object')

 "1.12. Crear la función para automatizar todo el proceso, todos los procesos anteriores que hemos ido realizando los encapsularemos dentro de una función. La función se llamará ``transform_data()``, el único requisito es que tenga el parámetro de la ruta donde se encuentran los datos. Utilizar la # para nombrar lo que se hace en cada parte del código.\n",

In [27]:
def transform_data(df):
  #Sustituir los valores Nan por 0
  df.fillna(0,inplace=True)
  #Eliminar la columna ``'Rk'`` y ``'matches'``.
  df.drop(["Rk","Matches"] ,axis=1,inplace=True)
  #Solucionar el problema de  los nombres de los jugadores
  df['Player']=df['Player'].apply(lambda name:name.split("\\")[0])
  #Solucionar el problema de la nacionalidad
  df['Nation']=df['Nation'].apply(lambda name:name.split(" ")[1])
  # Solucionar el problema de las posiciones
  df['Pos'] = df['Pos'].apply(lambda x: x[:2])
  #Renombrar las columnas que contiene el valor .1
  df.rename(columns={'Gls.1':'Gls/90','Ast.1':'Ast/90', 'G-PK.1':'G-PK/90','xG.1':'xG/90','xA.1':'xA/90','npxG.1':'npxG/90','npxG+xA.1':'npxG+xA/90'},inplace=True)
  return df

In [28]:
df_2=pd.read_csv("/content/drive/MyDrive/CURSO PYTHON/Tarea/fbref_laliga21 (3).csv")

In [29]:
df_2.head()

Rk                                     Player  Nation   Pos       Squad  \
0   1                Sabit Abdulai\Sabit-Abdulai  gh GHA    MF      Getafe   
1   2                  Marcos Acuña\Marcos-Acuna  ar ARG    DF     Sevilla   
2   3              Bobby Adekanye\Bobby-Adekanye  nl NED  FWMF       Cádiz   
3   4  Martin Agirregabiria\Martin-Agirregabiria  es ESP  DFMF      Alavés   
4   5                  Joseph Aidoo\Joseph-Aidoo  gh GHA    DF  Celta Vigo   

   Age  Born  MP  Starts   Min  ...   xG  npxG   xA  npxG+xA  xG.1  xA.1  \
0   21  1999   3       0    60  ...  0.0   0.0  0.0      0.0  0.00  0.00   
1   28  1991  30      26  2330  ...  1.1   1.1  2.8      4.0  0.04  0.11   
2   21  1999   3       0    36  ...  0.2   0.2  0.0      0.2  0.43  0.00   
3   24  1996  26      16  1558  ...  0.2   0.2  0.9      1.1  0.01  0.05   
4   24  1995  25      14  1289  ...  0.8   0.8  0.6      1.4  0.06  0.04   

   xG+xA  npxG.1  npxG+xA.1  Matches  
0   0.00    0.00       0.00  Matches  
1   0.15    0.04       0.15  Matches  
2   0.43    0.43       0.43  Matches  
3   0.06    0.01       0.06  Matches  
4   0.10    0.06       0.10  Matches  

[5 rows x 33 columns]

In [30]:
df_2=transform_data(df_2)
df_2.head()

Player Nation Pos       Squad  Age  Born  MP  Starts   Min  \
0         Sabit Abdulai    GHA  MF      Getafe   21  1999   3       0    60   
1          Marcos Acuña    ARG  DF     Sevilla   28  1991  30      26  2330   
2        Bobby Adekanye    NED  FW       Cádiz   21  1999   3       0    36   
3  Martin Agirregabiria    ESP  DF      Alavés   24  1996  26      16  1558   
4          Joseph Aidoo    GHA  DF  Celta Vigo   24  1995  25      14  1289   

    90s  ...  G+A-PK   xG  npxG   xA  npxG+xA  xG/90  xA/90  xG+xA  npxG/90  \
0   0.7  ...    0.00  0.0   0.0  0.0      0.0   0.00   0.00   0.00     0.00   
1  25.9  ...    0.12  1.1   1.1  2.8      4.0   0.04   0.11   0.15     0.04   
2   0.4  ...    0.00  0.2   0.2  0.0      0.2   0.43   0.00   0.43     0.43   
3  17.3  ...    0.06  0.2   0.2  0.9      1.1   0.01   0.05   0.06     0.01   
4  14.3  ...    0.00  0.8   0.8  0.6      1.4   0.06   0.04   0.10     0.06   

   npxG+xA/90  
0        0.00  
1        0.15  
2        0.43  
3        0.06  
4        0.10  

[5 rows x 31 columns]

2. Filtrado de datos
2.1 Con la técnica ``.loc[]``, seleccinamene todas las filas de las siguientes columnas:\n

In [31]:
cols=['Player', 'Squad', 'Gls/90', 'Ast/90', 'xG/90', 'xA/90']

In [32]:
df[cols]

Player          Squad  Gls/90  Ast/90  xG/90  xA/90
0           Sabit Abdulai         Getafe    0.00    0.00   0.00   0.00
1            Marcos Acuña        Sevilla    0.04    0.08   0.04   0.11
2          Bobby Adekanye          Cádiz    0.00    0.00   0.43   0.00
3    Martin Agirregabiria         Alavés    0.00    0.06   0.01   0.05
4            Joseph Aidoo     Celta Vigo    0.00    0.00   0.06   0.04
..                    ...            ...     ...     ...    ...    ...
577         Joseba Zaldúa  Real Sociedad    0.00    0.00   0.01   0.11
578          Oier Zarraga  Athletic Club    0.00    0.00   0.08   0.04
579         Igor Zubeldia  Real Sociedad    0.05    0.05   0.05   0.03
580      Martín Zubimendi  Real Sociedad    0.00    0.00   0.03   0.03
581       Martin Ødegaard    Real Madrid    0.00    0.00   0.04   0.15

[582 rows x 6 columns]

In [33]:
df.loc[:,cols]

Player          Squad  Gls/90  Ast/90  xG/90  xA/90
0           Sabit Abdulai         Getafe    0.00    0.00   0.00   0.00
1            Marcos Acuña        Sevilla    0.04    0.08   0.04   0.11
2          Bobby Adekanye          Cádiz    0.00    0.00   0.43   0.00
3    Martin Agirregabiria         Alavés    0.00    0.06   0.01   0.05
4            Joseph Aidoo     Celta Vigo    0.00    0.00   0.06   0.04
..                    ...            ...     ...     ...    ...    ...
577         Joseba Zaldúa  Real Sociedad    0.00    0.00   0.01   0.11
578          Oier Zarraga  Athletic Club    0.00    0.00   0.08   0.04
579         Igor Zubeldia  Real Sociedad    0.05    0.05   0.05   0.03
580      Martín Zubimendi  Real Sociedad    0.00    0.00   0.03   0.03
581       Martin Ødegaard    Real Madrid    0.00    0.00   0.04   0.15

[582 rows x 6 columns]

2.2. Obtén únicamente el nombre de los equipos.

In [44]:
df.loc[:,'Squad']

0             Getafe
1            Sevilla
2              Cádiz
3             Alavés
4         Celta Vigo
           ...      
577    Real Sociedad
578    Athletic Club
579    Real Sociedad
580    Real Sociedad
581      Real Madrid
Name: Squad, Length: 582, dtype: object

2.3. Nuevamente con ``.loc[]`` selecciona los datos del Atlético de madrid de la siguiente columnas.\n"

In [45]:
df.loc[df['Squad']=='Atlético Madrid',cols]

Player            Squad  Gls/90  Ast/90  xG/90  xA/90
92     Yannick Carrasco  Atlético Madrid    0.25    0.41   0.25   0.23
112        Ángel Correa  Atlético Madrid    0.34    0.30   0.29   0.21
115         Diego Costa  Atlético Madrid    0.87    0.00   0.60   0.14
124      Moussa Dembélé  Atlético Madrid    0.00    0.00   0.92   0.02
161              Felipe  Atlético Madrid    0.00    0.00   0.04   0.01
162          João Félix  Atlético Madrid    0.40    0.29   0.29   0.22
227       Mario Hermoso  Atlético Madrid    0.03    0.03   0.04   0.02
229      Héctor Herrera  Atlético Madrid    0.00    0.13   0.03   0.04
274                Koke  Atlético Madrid    0.03    0.06   0.06   0.08
276  Geoffrey Kondogbia  Atlético Madrid    0.00    0.00   0.02   0.04
291        Thomas Lemar  Atlético Madrid    0.07    0.20   0.09   0.13
296     Marcos Llorente  Atlético Madrid    0.37    0.33   0.13   0.16
297          Renan Lodi  Atlético Madrid    0.08    0.16   0.06   0.09
323  José María Giménez  Atlético Madrid    0.00    0.00   0.05   0.01
392         Saúl Ñíguez  Atlético Madrid    0.09    0.04   0.12   0.09
399           Jan Oblak  Atlético Madrid    0.00    0.00   0.00   0.00
417       Thomas Partey  Atlético Madrid    0.00    0.00   0.05   0.00
464              Ricard  Atlético Madrid    0.00    0.00   0.07   0.00
497      Manuel Sánchez  Atlético Madrid    0.00    1.84   0.00   0.61
501        Stefan Savić  Atlético Madrid    0.03    0.00   0.03   0.01
523         Luis Suárez  Atlético Madrid    0.75    0.11   0.56   0.12
533      Lucas Torreira  Atlético Madrid    0.17    0.00   0.04   0.10
540     Kieran Trippier  Atlético Madrid    0.00    0.22   0.00   0.13
566              Vitolo  Atlético Madrid    0.00    0.00   0.15   0.07
567       Šime Vrsaljko  Atlético Madrid    0.00    0.00   0.00   0.00

"2.4. Al filtrado anterior, añádele que me muestre también solo los delanteros del Atlético de Madrid con las columnas del ejercicio anterior."

In [38]:
cols_2=['Player','Pos', 'Squad', 'Gls/90', 'Ast/90', 'xG/90', 'xA/90']

In [39]:
df[cols_2]

Player Pos          Squad  Gls/90  Ast/90  xG/90  xA/90
0           Sabit Abdulai  MF         Getafe    0.00    0.00   0.00   0.00
1            Marcos Acuña  DF        Sevilla    0.04    0.08   0.04   0.11
2          Bobby Adekanye  FW          Cádiz    0.00    0.00   0.43   0.00
3    Martin Agirregabiria  DF         Alavés    0.00    0.06   0.01   0.05
4            Joseph Aidoo  DF     Celta Vigo    0.00    0.00   0.06   0.04
..                    ...  ..            ...     ...     ...    ...    ...
577         Joseba Zaldúa  DF  Real Sociedad    0.00    0.00   0.01   0.11
578          Oier Zarraga  MF  Athletic Club    0.00    0.00   0.08   0.04
579         Igor Zubeldia  DF  Real Sociedad    0.05    0.05   0.05   0.03
580      Martín Zubimendi  MF  Real Sociedad    0.00    0.00   0.03   0.03
581       Martin Ødegaard  MF    Real Madrid    0.00    0.00   0.04   0.15

[582 rows x 7 columns]

In [46]:
df.loc[(df['Pos']=='FW')&(df["Squad"]=='Atlético Madrid'),cols_2]

Player Pos            Squad  Gls/90  Ast/90  xG/90  xA/90
112    Ángel Correa  FW  Atlético Madrid    0.34    0.30   0.29   0.21
115     Diego Costa  FW  Atlético Madrid    0.87    0.00   0.60   0.14
124  Moussa Dembélé  FW  Atlético Madrid    0.00    0.00   0.92   0.02
162      João Félix  FW  Atlético Madrid    0.40    0.29   0.29   0.22
523     Luis Suárez  FW  Atlético Madrid    0.75    0.11   0.56   0.12

2.5. Utilizando  también  las columnas del ejercicio anterior, quiero seleccionar los delanteros del Real Madrid y Atlético de Madrid. Ordenar los datos en función de la columna ``'Gls'`` y de manera descendiente.

In [48]:
cols=['Player','Pos', 'Squad','Gls', 'Gls/90', 'Ast/90', 'xG/90', 'xA/90']

In [54]:
Delanteros=df.loc[(df['Pos']=='FW')&(df["Squad"]=='Atlético Madrid')|(df['Squad']=='Real Madrid')&(df['Pos']=='FW'),cols]

In [55]:
Delanteros

Player Pos            Squad  Gls  Gls/90  Ast/90  xG/90  xA/90
31    Sergio Arribas  FW      Real Madrid    0    0.00    0.00   0.02   0.03
33     Marco Asensio  FW      Real Madrid    5    0.24    0.10   0.23   0.15
55     Karim Benzema  FW      Real Madrid   23    0.72    0.28   0.55   0.19
112     Ángel Correa  FW  Atlético Madrid    9    0.34    0.30   0.29   0.21
115      Diego Costa  FW  Atlético Madrid    2    0.87    0.00   0.60   0.14
124   Moussa Dembélé  FW  Atlético Madrid    0    0.00    0.00   0.92   0.02
143        Hugo Duro  FW      Real Madrid    0    0.00    0.00   0.22   0.09
162       João Félix  FW  Atlético Madrid    7    0.40    0.29   0.29   0.22
226      Eden Hazard  FW      Real Madrid    3    0.51    0.34   0.27   0.14
261       Luka Jović  FW      Real Madrid    0    0.00    0.00   0.67   0.08
263  Vinicius Júnior  FW      Real Madrid    3    0.14    0.14   0.30   0.14
324          Mariano  FW      Real Madrid    1    0.18    0.00   0.26   0.03
335    Borja Mayoral  FW      Real Madrid    0    0.00    0.00   0.00   0.00
480          Rodrygo  FW      Real Madrid    1    0.09    0.55   0.23   0.28
523      Luis Suárez  FW  Atlético Madrid   21    0.75    0.11   0.56   0.12

In [56]:
Delanteros.sort_values('Gls',ascending=False)

Player Pos            Squad  Gls  Gls/90  Ast/90  xG/90  xA/90
55     Karim Benzema  FW      Real Madrid   23    0.72    0.28   0.55   0.19
523      Luis Suárez  FW  Atlético Madrid   21    0.75    0.11   0.56   0.12
112     Ángel Correa  FW  Atlético Madrid    9    0.34    0.30   0.29   0.21
162       João Félix  FW  Atlético Madrid    7    0.40    0.29   0.29   0.22
33     Marco Asensio  FW      Real Madrid    5    0.24    0.10   0.23   0.15
226      Eden Hazard  FW      Real Madrid    3    0.51    0.34   0.27   0.14
263  Vinicius Júnior  FW      Real Madrid    3    0.14    0.14   0.30   0.14
115      Diego Costa  FW  Atlético Madrid    2    0.87    0.00   0.60   0.14
324          Mariano  FW      Real Madrid    1    0.18    0.00   0.26   0.03
480          Rodrygo  FW      Real Madrid    1    0.09    0.55   0.23   0.28
31    Sergio Arribas  FW      Real Madrid    0    0.00    0.00   0.02   0.03
124   Moussa Dembélé  FW  Atlético Madrid    0    0.00    0.00   0.92   0.02
143        Hugo Duro  FW      Real Madrid    0    0.00    0.00   0.22   0.09
261       Luka Jović  FW      Real Madrid    0    0.00    0.00   0.67   0.08
335    Borja Mayoral  FW      Real Madrid    0    0.00    0.00   0.00   0.00

2.6. Realizar una agrupación por los equipos, de las columnas ``'Gls', 'Ast', 'xG', 'xA'``\n",
        "y calcula el promedio y el total para cada columna.\n"

In [57]:
df.columns

Index(['Player', 'Nation', 'Pos', 'Squad', 'Age', 'Born', 'MP', 'Starts',
       'Min', '90s', 'Gls', 'Ast', 'G-PK', 'PK', 'PKatt', 'CrdY', 'CrdR',
       'Gls/90', 'Ast/90', 'G+A', 'G-PK/90', 'G+A-PK', 'xG', 'npxG', 'xA',
       'npxG+xA', 'xG/90', 'xA/90', 'xG+xA', 'npxG/90', 'npxG+xA/90'],
      dtype='object')

In [65]:
df.groupby('Squad').agg({'Gls':['mean','sum'],'Ast':['mean','sum'],'xG':['mean','sum'],'xA':['mean','sum']})

Gls           Ast            xG              xA      
                     mean sum      mean sum      mean   sum      mean   sum
Squad                                                                      
Alavés           1.166667  35  0.700000  21  1.440000  43.2  0.860000  25.8
Athletic Club    1.592593  43  1.148148  31  1.659259  44.8  1.148148  31.0
Atlético Madrid  2.600000  65  2.160000  54  2.112000  52.8  1.516000  37.9
Barcelona        3.200000  80  2.040000  51  3.208000  80.2  2.304000  57.6
Betis            2.000000  50  1.400000  35  1.896000  47.4  1.200000  30.0
Celta Vigo       1.833333  55  1.300000  39  1.610000  48.3  1.120000  33.6
Cádiz            0.970588  33  0.470588  16  1.026471  34.9  0.564706  19.2
Eibar            0.966667  29  0.666667  20  1.416667  42.5  0.880000  26.4
Elche            1.100000  33  0.833333  25  1.023333  30.7  0.730000  21.9
Getafe           0.870968  27  0.451613  14  1.058065  32.8  0.651613  20.2
Granada          1.352941  46  0.882353  30  1.176471  40.0  0.700000  23.8
Huesca           1.185185  32  0.851852  23  1.322222  35.7  0.885185  23.9
Levante          1.551724  45  1.206897  35  1.413793  41.0  0.955172  27.7
Osasuna          1.285714  36  0.892857  25  1.367857  38.3  0.835714  23.4
Real Madrid      2.133333  64  1.766667  53  2.090000  62.7  1.540000  46.2
Real Sociedad    1.933333  58  1.300000  39  2.050000  61.5  1.310000  39.3
Sevilla          2.000000  52  1.384615  36  1.980769  51.5  1.330769  34.6
Valencia         1.600000  48  1.166667  35  1.480000  44.4  0.903333  27.1
Valladolid       1.062500  34  0.656250  21  1.143750  36.6  0.693750  22.2
Villarreal       1.965517  57  1.137931  33  2.017241  58.5  1.400000  40.6

3. Bucle for y condicionales\n
3.1. Vamos a realizar un bucle for donde iteraremos por los diferentes valores de la columna ``'Gls'`` para crear una nueva columna que se llamara ``'type_goal_player'``. Dejo a tu elección los valores en base dónde cortar para crear las diferentes categorías. Hay que crear tres valor Poco goleador, goleador, Muy goleador\n"

In [76]:
for i in range(len(df)):

  if df.Gls.values[i] <= 5:
    df.loc[i,'type_goal_player'] = 'Poco goleador'

  elif (df.Gls.values[i]  > 5) & (df.Gls.values[i]  <= 15):
    df.loc[i,'type_goal_player']= 'Goleador'

  else :
    df.loc[i,'type_goal_player'] = 'Muy goleador'



3.2. Filtramos los datos para los jugadores que tienen el valor de ``'Muy goleador'`` en nuestra nueva columna."

In [81]:
df[df['type_goal_player']=='Muy goleador']

Player Nation Pos            Squad  Age  Born  MP  Starts  \
55       Karim Benzema    FRA  FW      Real Madrid   32  1987  34      33   
147  Youssef En-Nesyri    MAR  FW          Sevilla   23  1997  38      23   
242     Alexander Isak    SWE  FW    Real Sociedad   20  1999  34      30   
344       Lionel Messi    ARG  FW        Barcelona   33  1987  35      33   
369      Gerard Moreno    ESP  FW       Villarreal   28  1992  33      30   
523        Luis Suárez    URU  FW  Atlético Madrid   33  1987  32      30   

      Min   90s  ...    xG  npxG    xA  npxG+xA  xG/90  xA/90  xG+xA  npxG/90  \
55   2894  32.2  ...  17.8  17.0   6.3     23.3   0.55   0.19   0.75     0.53   
147  2311  25.7  ...  14.8  14.8   1.3     16.1   0.58   0.05   0.63     0.58   
242  2340  26.0  ...  15.5  14.7   2.3     17.0   0.60   0.09   0.68     0.57   
344  3023  33.6  ...  23.6  19.4  10.8     30.3   0.70   0.32   1.02     0.58   
369  2673  29.7  ...  19.1  11.6   7.6     19.2   0.64   0.26   0.90     0.39   
523  2508  27.9  ...  15.6  13.3   3.3     16.6   0.56   0.12   0.68     0.48   

     npxG+xA/90  type_goal_player  
55         0.72      Muy goleador  
147        0.63      Muy goleador  
242        0.66      Muy goleador  
344        0.90      Muy goleador  
369        0.65      Muy goleador  
523        0.59      Muy goleador  

[6 rows x 32 columns]

3.3. Quiero saber la cantidad de jugadores de cada tipo de goleador. Cuantas veces sale cada valor en la columna ``'type_goal_player'``.

In [78]:
df.type_goal_player.value_counts()

Poco goleador    541
Goleador          35
Muy goleador       6
Name: type_goal_player, dtype: int64